rust/connlib/tunnel/src/client.rs

```diff
@@ -400,31 +387,9 @@ impl ClientState {
             .get_mut(&resource)
             .ok_or(Error::UnexpectedConnectionDetails)?;
 
-        if self.gateway_awaiting_connection.contains(&gateway) {
-            self.awaiting_connection.remove(&resource);
-            return Err(Error::PendingConnection);
-        }
```

Comment:
Yeah? If we don't have a connection, we send an intent unless we've sent one in the last 2 seconds. If we do have a connection, we use it. I don't really see how we could make multiple.

Reply:
We can have multiple resource that have an intent to the same gateway, each of them sends a connection intent with a different resource id but all come back with the same gateway id.

We'd have no way to keep track that we are trying to connect to the same gateway for multiple difference resources since we only keep track of connection intents by resource id.

Reply:
Okay, well how do we want to solve this then? Because if I add this condition back, connections don't work at all!

Wouldn't it be cleaner to check for an existing connection to that gateway when we accept one instead of replacing it?

Reply:
The problem with that is that there's no way to distinguish, with the control-plane as it is at least, between a client that restarted and a new connection intent from the same client.

I'm not sure why it didn't work with the `gateway_awaiting_connection` we're probably forgetting to clear it somewhere.

That doesn't mean that I'm not open to changing how this works.

Reply:
> The problem with that is that there's no way to distinguish, with the control-plane as it is at least, between a client that restarted and a new connection intent from the same client.

We are on the client here, not the gateway so I am not sure what you mean by this.

Reply:
> Wouldn't it be cleaner to check for an existing connection to that gateway when we accept one instead of replacing it?

Can you explain why this wouldn't work? Even if we emit multiple intents to different resources that all end up making us connect to the same gateway, we can simply _not_ override the state of the connection but just extend it, no? Wouldn't we just emit `ReuseConnection` for the ones following the first one?

Reply:
I think the state that is mis-aligned here is that we make a connection using `snownet` but we don't create a `Peer` entry yet. What we need is an additional check whether we are currently connecting to the gateway and if yes, we return `ReuseConnection`.

Reply:
> We are on the client here, not the gateway so I am not sure what you mean by this.

Basically, when we make the connection the gateway will receive a client id, because we send a `RequestConnection` if this is a restarted client, we want the connection to be restarted from the gateway perspective. So we make sure to send the message once per-client boot with `gateway_awaiting_connection` or with the peer itself after being connected.

> Can you explain why this wouldn't work? Even if we emit multiple intents to different resources that all end up making us connect to the same gateway, we can simply not override the state of the connection but just extend it, no? Wouldn't we just emit ReuseConnection for the ones following the first one?

For this to work we would need to buffer `ReuseConnection` on the gateway, which would be fine, as long as we make sure to bound the buffer. But to not do that we used the `gateway_awaiting_connection`. We need to do it because we can't make sure of the order of message arrival in the gateway.

> I think the state that is mis-aligned here is that we make a connection using snownet but we don't create a Peer entry yet. What we need is an additional check whether we are currently connecting to the gateway and if yes, we return ReuseConnection.

Yeah, we would need to do that

Reply:
Let me know what you think of the last commit, doesn't that also fix it? I think it is conceptually the same just with a different condition.

Reply:
yeah, I think this works :)